hudi-spark-datasource/hudi-spark/src/test/scala/org/apache/hudi/common/table/read/TestSpark35RecordPositionMetadataColumn.scala

```diff
@@ -0,0 +1,112 @@
+  private val HOOD_FORMAT = "hudi"
+  private val PARQUET_FORMAT = "parquet"
+  private val ROW_INDEX_COLUMN = "_tmp_metadata_row_index"
+  private val SPARK_MERGER = "org.apache.hudi.HoodieSparkRecordMerger"
+
+  @Test
+  def testRecordPositionColumn(): Unit = {
+    val _spark = spark
+    import _spark.implicits._
+
+    val userToCountryDF = Seq(
+      (1, "US", "1000"),
+      (2, "US", "1000"),
+      (3, "China", "1000"),
+      (4, "Singapore", "1000")
+    ).toDF("userid", "country", "ts")
+    val userToCountryTblPath = s"$basePath/user_to_country_table"
+
+    userToCountryDF.write.format(HOOD_FORMAT)
+      .option(DataSourceWriteOptions.RECORDKEY_FIELD.key, "userid")
+      .option(DataSourceWriteOptions.PRECOMBINE_FIELD.key, "ts")
+      .option(HoodieWriteConfig.TBL_NAME.key, "user_to_country")
+      .option(HoodieWriteConfig.RECORD_MERGER_IMPLS.key, SPARK_MERGER)
+      .option(HoodieWriteConfig.WRITE_RECORD_POSITIONS.key, "true")
+      .option(HoodieStorageConfig.LOGFILE_DATA_BLOCK_FORMAT.key, PARQUET_FORMAT)
+      .option(DataSourceWriteOptions.TABLE_TYPE.key(), HoodieTableType.MERGE_ON_READ.name())
+      .save(userToCountryTblPath)
+
+    val dataSchema = new StructType(
+      Array(
+        StructField("userid", IntegerType, nullable = false),
+        StructField("country", StringType, nullable = false),
+        StructField("ts", StringType, nullable = false)
+      )
+    )
+    val requiredSchema = HoodieFileGroupReaderBasedParquetFileFormat
+      .getAppliedRequiredSchema(
+        dataSchema,
+        shouldUseRecordPosition = true,
+        ROW_INDEX_COLUMN)
+    if (HoodieSparkUtils.gteqSpark3_5) {
+      assertEquals(4, requiredSchema.fields.length)
+      assertEquals(
+        "StructField(_tmp_metadata_row_index,LongType,false)",
+        requiredSchema.fields(3).toString)
+    }
+
+    _spark.conf.set("spark.sql.parquet.enableVectorizedReader", "false")
+    val metaClient = getHoodieMetaClient(
+      spark().sparkContext.hadoopConfiguration, userToCountryTblPath)
+    val fileReader = new ParquetFileFormat().buildReaderWithPartitionValues(
+      spark(),
+      dataSchema,
+      StructType(Nil),
+      requiredSchema,
+      Nil,
+      Map.empty,
+      new Configuration(spark().sparkContext.hadoopConfiguration))
+    val allBaseFiles = HoodieTestTable.of(metaClient).listAllBaseFiles
+    assertTrue(allBaseFiles.nonEmpty)
+
+    if (HoodieSparkUtils.gteqSpark3_5) {
+      val fileInfo = sparkAdapter.getSparkPartitionedFileUtils
+        .createPartitionedFile(InternalRow.empty, allBaseFiles.head.getPath, 0, allBaseFiles.head.getLen)
+      val iterator = new CloseableInternalRowIterator(fileReader.apply(fileInfo))
+      var rowIndices: Set[Long] = Set()
+      while (iterator.hasNext) {
+        val row = iterator.next()
+        rowIndices += row.getLong(3)
+      }
+      val expectedRowIndices: Set[Long] = Set(0L, 1L, 2L, 3L)
```

Comment:
Could you add another test case where the some rows are filtered out (e.g., 0 and 1 are filtered out) and the row positions returned are still correct?